crates/types/src/epoch.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+#![allow(dead_code)]
+
+use crate::identifiers::{LeaderEpoch, PartitionId};
+use crate::{GenerationalNodeId, Version, Versioned};
+
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct Epoch {
+    version: Version,
+
+    // Metadata information about the last known leader
+    partition_id: PartitionId,
+    node_id: GenerationalNodeId,
+}
+
+impl Versioned for Epoch {
+    fn version(&self) -> Version {
+        self.version
+    }
+}
+
+impl Epoch {
+    pub fn new(node_id: GenerationalNodeId, partition_id: PartitionId) -> Self {
+        Self {
+            version: Version::MIN,
+            node_id,
+            partition_id,
+        }
+    }
+
+    pub fn epoch(&self) -> LeaderEpoch {
+        // todo think about aligning Version and LeaderEpoch types
```

Comment:
Maybe make use of types/src/logs.rs (trait SequenceNumber) instead?

Reply:
True, this could also be a solution. The conversion `Version -> u32 -> u64 -> LeaderEpoch` is not super nice but also not incorrect at the moment.